From 9727d2dbcfe059d12e0df73fcc28a67b1171dfe4 Mon Sep 17 00:00:00 2001
From: John Johansen <john.johansen@canonical.com>
Date: Tue, 6 Oct 2015 17:00:35 -0700
Subject: [PATCH 0857/1767] UBUNTU: SAUCE: apparmor: Fix: label_vec_merge
 insertion

BugLink: http://bugs.launchpad.net/bugs/1448912

label_vec_merge should only do the insertion after the vector is copied.

Signed-off-by: John Johansen <john.johansen@canonical.com>
Signed-off-by: Tim Gardner <tim.gardner@canonical.com>
---
 security/apparmor/label.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/security/apparmor/label.c b/security/apparmor/label.c
index 218d856..627109a 100644
--- a/security/apparmor/label.c
+++ b/security/apparmor/label.c
@@ -1160,11 +1160,11 @@ struct aa_label *aa_label_vec_merge(struct aa_profile **vec, int len,
 	if (!new)
 		return NULL;
 
-	write_lock_irqsave(&ls->lock, flags);
 	for (i = 0; i < len; i++) {
 		new->ent[i] = aa_get_profile(vec[i]);
-		label = __aa_label_insert(ls, new, false);
 	}
+	write_lock_irqsave(&ls->lock, flags);
+	label = __aa_label_insert(ls, new, false);
 	write_unlock_irqrestore(&ls->lock, flags);
 	if (label != new)
 		/* not fully constructed don't put */
-- 
1.9.3

